packages/net/http/_test.pony

```diff
@@ -362,3 +366,192 @@ primitive _Test
     h.assert_eq[String](path, url.path)
     h.assert_eq[String](query, url.query)
     h.assert_eq[String](fragment, url.fragment)
+
+// Actor and classes to test the HTTPClient and modified _HTTPConnection.
+actor _HTTPHandlerActor
+  let h: TestHelper
+  var tries: USize
+  
+  new create(h': TestHelper, tries': USize) =>
+    h = h'
+    tries = tries'
+    h.log("_HTTPHandlerActor create called")
+
+  be apply(p: Payload val) =>
+    h.log("_HTTPHandlerActor apply called. Tries to go: " + tries.string())
+    if (tries = tries - 1) == 1 then 
+      h.complete(true)
+    end
+
+class _TestHTTPHandler is HTTPHandler
+  let h: TestHelper
+  let ha: _HTTPHandlerActor
+
+  new create(ha': _HTTPHandlerActor, h': TestHelper) =>
+    ha = ha'
+    h = h'
+    h.log("_TestHTTPHandler.create called")
+
+  fun ref apply(payload: Payload val): Any => 
+    h.log("_TestHTTPHandler.apply called")
+    ha(payload)
+
+  fun ref chunk(data: ByteSeq val) => 
+    h.log("_TestHTTPHandler.chunk called")
+
+class val _TestHandlerFactory is HandlerFactory
+  let h: TestHelper
+  let ha: _HTTPHandlerActor
+
+  new val create(ha': _HTTPHandlerActor, h': TestHelper) =>
+    ha = ha'
+    h = h'
+
+  fun apply(session: HTTPSession): HTTPHandler ref^ =>
+    h.log("_TestHTTPHandlerFactory.apply called")
+    _TestHTTPHandler(ha, h)
+
+class iso _HTTPConnTest is UnitTest
+  var server: (TCPListener | None) = None
+  fun name(): String => "net/http/_HTTPConnection._new_conn"
+  fun label(): String => "conn-fix"
+
+  fun ref apply(h: TestHelper) ? =>
+    let urls: Array[URL] = 
+      [ 
+        URL.build(
```

Comment:
testing the connection once one should be enough, right?

Reply:
It is essential that the request is made more than once.

This is the flow of the old code before the fix:

1. be apply is called with payload 1
2. apply pushes payload 1 on the _unsent list
3. _send_pending is called
1. _send_pending tries to cast _conn as TCPConnection and fails, because _conn is None; it calls _new_conn()
1. _new_conn creates a new TCPConnection and assigns it to _conn
1. be apply returns
1. be apply is called with payload 2
1. apply pushes the payload on the _unsent list which now contains two payloads
1. apply calls _send_pending
1. _send_pending send_pending tries to cast _conn as TCPConnection and succeeds, but _conn is not open yet
1. _send_pending continues to shift the payloads from the _unsent list and calls request.write, which it thinks succeeds.
1. when all the request have been processed _send_pending and apply return and some time after that the be _connected behaviour is called and now _conn is really open and ready for use.

It works when you only send one payload, because _send_pending is called again after be _connected is called and _conn is a connected TCPConnection.

This is the scenario:

1. be apply is called with payload 1
2. apply pushes payload 1 on the _unsent list
1. _send_pending is called
1. _send_pending tries to cast _conn as TCPConnection and fails, because _conn is None; it calls _new_conn()
1. _new_conn creates a new TCPConnection and assigns it to _conn
1. be _connected is called and calls _send_pending
1. _send_pending works fine now, because _conn is actually connected.

Reply:
Totally true. You need to send at least two requests.
But it would suffice to create 1 url, as it is always the same and just use a `collections.Range` from 0 to 5 or so to iterate over and send a request on each iteration.

Reply:
That is true, but in my initial test I sent multiple request to different sites and this code made it easier to perform these tests. 

I kept it in because there are more bugs in the http package and I may need it later. But I can take it out if you like.

Reply:
Just to verify. Will the CI start a new build when I push new commits to the PR branch?

Reply:
One idea still. Sorry if i seem picky but i did this wrong too many times in the past.

If you bind to port 50000 it might be already taken on someones machine and thus the testsuite might fail although pony is just fine.

My suggestion is to bind to port 0 with the `TCPListener` and to get the randomly assigned free port back in the `TCPListenNotify.connected` method from `TCPListener.local_address().port`. This way you make sure that you will most likely always get a free port and the test suite runs just fine.

Reply:
Afaik the Ci will build every commit you make to a branch that has a PR open in the ponc repo

Reply:
Will do that. Good tip. Did not know the feature yet.

Reply:
I assume that you mean the TCPListenNotify.listening method.
The TCPListener.local_address().port returns a strange value. When I set the port value to e.g. 50000, then the port has value 20675, when I set it to 5000 port has value 34835. I have no idea why this is. But I assume it is undesired behaviour.

Reply:
i sure mean `TCPListenNotify.listening`.
sorry for the confusion here.

The port from `net.NetAddress` is actually from the OS `sockaddr_in` (`sockaddr_in6`) structure defined in C. This is an unsigned short (`U16`) in network byte order. I learned that right now, sorry for that. :)
Here you have a playground link that shows this: 

http://playground.ponylang.org/?gist=f783b91faedef2715d4a585f59ec4039

looking forward to merge your PR.